monitoring/v1/types.go

```diff
@@ -0,0 +1,349 @@
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	// items is a list of AlertRelabelConfigs.
+	Items []*AlertRelabelConfig `json:"items"`
+}
+
+// LabelName is a valid Prometheus label name which may only contain ASCII
+// letters, numbers, and underscores.
+//
+// +kubebuilder:validation:Pattern:="^[a-zA-Z_][a-zA-Z0-9_]*$"
+type LabelName string
+
+// RelabelConfig allows dynamic rewriting of label sets for alerts.
+// See Prometheus documentation:
+// - https://prometheus.io/docs/prometheus/latest/configuration/configuration/#alert_relabel_configs
+// - https://prometheus.io/docs/prometheus/latest/configuration/configuration/#relabel_config
+//
+// +k8s:openapi-gen=true
+type RelabelConfig struct {
+	// sourceLabels select values from existing labels. Their content is
+	// concatenated using the configured separator and matched against the
+	// configured regular expression for the Replace, Keep, and Drop actions.
+	//
+	// +optional
+	// +kubebuilder:validation:XValidation:rule="self.action != 'Replace' && self.action != 'Keep' && self.action != 'Drop' && self.action != 'HashMod' || size(self.sourceLabels) > 0",message="sourceLabels is required for actions Replace, Keep, Drop and HashMod"
+	SourceLabels []LabelName `json:"sourceLabels,omitempty"`
+
+	// separator placed between concatenated source label values. When omitted,
+	// Prometheus will use its default value of ';'.
+	//
+	// +optional
+	Separator string `json:"separator,omitempty"`
+
+	// targetLabel to which the resulting value is written in a 'Replace' action.
+	// It is mandatory for 'Replace' and 'HashMod' actions. Regex capture groups
+	// are available.
+	//
+	// +optional
+	// +kubebuilder:validation:XValidation:rule="self.action != 'Replace' && self.action != 'HashMod' || has(self.targetLabel)",message="targetLabel is required when action is Replace or HashMod"
+	TargetLabel string `json:"targetLabel,omitempty"`
+
+	// regex against which the extracted value is matched. Default is: '(.*)'
+	// regex is required for all actions except 'HashMod'
+	//
+	// +optional
+	// +kubebuilder:validation:XValidation:rule="self.action != 'HashMod' || has(self.regex)",message="regex is required for all action expect HashMod"
+	Regex string `json:"regex,omitempty"`
+
+	// modulus to take of the hash of the source label values.  This can be
+	// combined with the 'HashMod' action to set 'target_label' to the 'modulus'
+	// of a hash of the concatenated 'source_labels'. This is only valid of sourceLabels is not empty
+	//
+	// +optional
+	// +kubebuilder:validation:XValidation:rule="!has(self.sourceLabels) || has(self.modulus) && size(self.sourceLabels) > 0",message="modulus requires sourceLabels to be present"
+	Modulus uint64 `json:"modulus,omitempty"`
+
+	// replacement value against which a regex replace is performed if the regular
+	// expression matches. This is required if the action is 'Replace' or
+	// 'LabelMap'. Regex capture groups are available. Default is: '$1'
```

Comment:
also valid for Keep, Drop, LabelDrop and LabelKeep actions?

Reply:
Valid yes, but its not required for those actions iiuc.

Reply:
Same as for "replace" and "labelmap" actions I believe:  Prometheus will default to "$1" so users can omit the "replacement" value if the default value is ok for them.

Reply:
So, this only works with a subset of the actions? And in some it's required, some it's defaulted? Or are they all behaving the same?

Reply:
I dropped the kubebuilder default annotation since for some actions this and other fields are required to be empty. For the actions that don't require a `replacement` prometheus itself will set the default if needed. For the actions that require `replacement` I added a cross validation rule.

Reply:
Is it optional or not allowed for others? We should specify within the godoc

Either `and optional for other actions`, or `and forbidden for other actions`